### ISSUE TITLE
Make mocks movable and copyable

### ISSUE THREAD

**Issue by jdfa** [google/googletest#3734]
**Does the feature exist in the most recent commit?**

No

**Why do we need this feature?**

Currently mock objects can't be used when tested class is calling copy or move on them. All of these `static_assert` are failing now:
```cpp 
class MyClassMock {
 public:
  MOCK_METHOD(void, MyFunction, (int));
};

static_assert(std::is_move_assignable_v<MyClassMock>, "Mock move assignable");
static_assert(std::is_move_constructible_v<MyClassMock>,
              "Mock move constructible");
static_assert(std::is_copy_assignable_v<MyClassMock>, "Mock copy assignable");
static_assert(std::is_copy_constructible_v<MyClassMock>,
              "Mock copy constructible");

```

That's limiting testing options, especially for templated classes.

**Describe the proposal**

Consider this case, testing of simple templated class which will copy input parameter:
```cpp
class Tool {
 public:
  void DoJob() const {}
};

class ToolMock {
 public:
  MOCK_METHOD(void, DoJob, (), (const));
};

template <typename TTool>
class ToolUser {
 public:
  ToolUser(const TTool& i_tool) : m_tool(i_tool) {}

  void UseTool() { m_tool.DoJob(); }

 private:
  TTool m_tool;
};

TEST(ToolUser, ShouldCallDoJob_FromTool) {
  ToolMock tool_mock;
  EXPECT_CALL(tool_mock, DoJob).Times(1);

  ToolUser<ToolMock> user(tool_mock);
  user.UseTool();
}
```
Currently it is failing to compile ToolUser ctor.

I imagine that it should be possible to make this work with something alike reference counting for mocked classes.  
Not sure about api for this feature, probably it will depend on what implementation is possible.  
Also I think there must be an option to not populate expectations on copied/moved objects.

**Is the feature specific to an operating system, compiler, or build system version?**

No

**Comment by sbenzaquen** [comment]
This limitation is not about implementation details, but about API design.
If the type were to be copyable, then the operation has to have clear and obvious behavior.
What should that behavior be? If users can't immediately know what it is without reading documentation then it might be better to not have such a feature.

Should the copy share the expectations? If so, you can get to race conditions and other issues between the copies.
Should the copy make its own copy of the expectations? If so, then you are duplicating expectations. You might have said `Times(1).WillOnce(do_x)` and due to copying you might have `do_x` happen more than once.
Should the copy have no expectations transferred? If so, your test case above would not work.

**Comment by jdfa** [comment]
Agree, probably request must be titled as "Make possibility to track calls of copied/moved mocks". In that case default behavior will be the same but new behavior will be enabled which specific action.  
I don't think that syntax alike NiceMock/StrictMock templates is appropriate here since copyable and movable is a property of class, not an object. I would prefer some statement in class itself.

I can definitely imagine that there are scenarios when you just want to track method call from any object (original one or it's copy), like in scenario above. So if we have expectation `Times(1)` - then in will be fulfilled if either method will be called of original object or from it's copy (or from copy of copy).
Also there might be scenario when you want to use mock in copy-friendly environment but don't want to track method calls from copies. 

Regarding move it's similar situation: populate expectations of original mock object to all objects constructed from rvalue-ref or  just allow to use mock object in move-friendly environment while tracking expectations only for original one.

**Comment by derekmauro** [comment]
Thank you for the suggestion. We aren't going to consider doing this for the reasons @sbenzaquen discussed.

**Comment by AndersSundman** [comment]
I would ask you to please re-consider this feature request. Pretty please 🙏 ❤️ 

TL;DR version:

With the introduction of concepts and CTAD in C++, together with the growing realization of just how awesome value semantics are, dependency injection using templated, and concept conforming values, are becoming much more prevalent. I would go as far as to say it's the new normal, unless you really need runtime polymorphism. It's a shame that there is no easy way to test this style of code using gtest/gmock.

The slightly longer version:

I realize that this is an API design issue. Providing mocking of copyable objects requires clarity on the mocking semantics. Perhaps you could provide some way of opting in to this by providing some kind of configuration of the mock object when it's defined? This would be my suggested semantics:
* Make the mock movable. Since there is only ever one valid mock object, that's the one that should record actions and do validation. The moved from husks should not record any further actions and not do any validation.
* Make the mock copyable and treat all copies like shared_ptr to one and the same mock. I.e. all objects record actions but only the last one standing does validation.

Here is an example of code that would be awesome if it was possible to get working (without having to resort to pointers and dynamic memory allocations):

```c++
#include <concepts>
#include <memory>

template <typename T>
concept Engine = requires(T c) {
                     { c.start() } -> std::convertible_to<bool>;
                 };

template <Engine TEngine>
struct Car {
    explicit Car(const TEngine& engine) : engine(engine) {}
    explicit Car(TEngine&& engine) : engine(std::move(engine)) {}

    bool start() { return engine.start(); }

   private:
    TEngine engine;
};

struct ElectricEngine {
    bool start() { return true; }
};
static_assert(Engine<ElectricEngine>);


// Test code 

#include "gmock/gmock.h"
#include "gtest/gtest.h"
using ::testing::Return;

struct MockEngine {
   public:
    MOCK_METHOD(bool, start, (), ());
};
static_assert(Engine<MockEngine>);


TEST(car_tests, engine_start) {
    // This is similar to the production code
    Car electric_car{ElectricEngine{}};
    ASSERT_TRUE(electric_car.start());

    // This would be nice to be able to do
    MockEngine mock_engine{};
    EXPECT_CALL(mock_engine, start()).Times(1).WillOnce(Return(true));

    // Car car{std::move(mock_engine)};
    // or Car car{mock_engine};
    // ASSERT_TRUE(car.start());
}
```

**Comment by AndersSundman** [comment]
:point_up: @derekmauro @sbenzaquen 

Sorry for pinging you if you already saw this. I just didn't want it to go unnoticed since it was a closed issue.
